### PR TITLE
'uart' and 'improv_serial' need to understand non-UART logger configurations

### PR DESCRIPTION
**Pull request by kpfleming**
# What does this implement/fix?

Ensures that all hardware UARTs can be used when the logger is using USB_CDC or USB_SERIAL_JTAG. Also adds component tests for `logger` configurations using those interfaces, and adds tests for `uart` configurations with the maximum number of hardware UARTs along with the logger using non-UART interfaces (those tests cannot currently fail as the 'maximum number' check is performed at run-time, not configuration-time or compile-time).

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

### CODE REVIEW

**Review by jesserockz — APPROVED** [review]

File: esphome/components/improv_serial/improv_serial_component.cpp [hunk 1]
```diff
@@ -99,19 +99,19 @@ void ImprovSerialComponent::write_data_(std::vector<uint8_t> &data) {
 #endif  // !USE_ESP32_VARIANT_ESP32C3 && !USE_ESP32_VARIANT_ESP32S2 && !USE_ESP32_VARIANT_ESP32S3
       uart_write_bytes(this->uart_num_, data.data(), data.size());
       break;
-#if defined(CONFIG_ESP_CONSOLE_USB_CDC) && (defined(USE_ESP32_VARIANT_ESP32S2) || defined(USE_ESP32_VARIANT_ESP32S3))
+#ifdef USE_LOGGER_USB_CDC
```
Comment (kbx81) [inline esphome/components/improv_serial/improv_serial_component.cpp:102]: So I'm just realizing that, when this is set but `CONFIG_ESP_CONSOLE_USB_CDC` is not set, some code is excluded by IDF from the build, resulting in a linker error...

Comment (kbx81) [reply esphome/components/improv_serial/improv_serial_component.cpp:102]: Specifically, I believe this is the point of interest: https://github.com/espressif/esp-idf/blob/38eeba213aa695aabfd6d89aa9f5078dbe5a94c3/components/esp_system/port/soc/esp32s3/CMakeLists.txt#L19-L21
...so I think that's not being built, but we are still including code based on this define, which results in the linker error.

Comment (kbx81) [reply esphome/components/improv_serial/improv_serial_component.cpp:102]: I think the right move might be to do this:
```suggestion
#if defined(USE_LOGGER_USB_CDC) && defined(CONFIG_ESP_CONSOLE_USB_CDC)
```
@kpfleming any thoughts?

Comment (kbx81) [reply esphome/components/improv_serial/improv_serial_component.cpp:102]: We also need to add more Improv serial tests for each ESP that natively supports USB -- it probably would've caught the issue. 🤦🏻

Comment (kbx81) [reply esphome/components/improv_serial/improv_serial_component.cpp:102]: The linker error can be seen [here](https://github.com/esphome/esphome/actions/runs/9925278247/job/27417528461?pr=7081).
PR to fix is #7082 -- we can continue discussion there, if necessary.

**Pull request merged**